google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2.java

```diff
@@ -0,0 +1,364 @@
+                + streamName
+                + " Error: "
+                + e.toString());
+      } finally {
+        this.lock.unlock();
+      }
+
+      if (localQueue.isEmpty()) {
+        continue;
+      }
+
+      // TODO: Add reconnection here.
+      while (!localQueue.isEmpty()) {
+        this.streamConnection.send(localQueue.pollFirst().message);
+      }
+    }
+
+    log.info("Cleanup starts. Stream: " + streamName);
+    // At this point, the waiting queue is drained, so no more requests.
+    // We can close the stream connection and handle the remaining inflight requests.
+    this.streamConnection.close();
+    waitForDoneCallback();
+
+    // At this point, there cannot be more callback. It is safe to clean up all inflight requests.
+    log.info(
+        "Stream connection is fully closed. Cleaning up inflight requests. Stream: " + streamName);
+    cleanupInflightRequests();
+    log.info("Append thread is done. Stream: " + streamName);
+  }
+
+  /*
+   * Returns true if waiting queue is drain, a.k.a. no more requests in the waiting queue.
+   *
+   * It serves as a signal to append thread that there cannot be any more requests in the waiting
+   * queue and it can prepare to stop.
+   */
+  private boolean waitingQueueDrained() {
+    this.lock.lock();
+    try {
+      return (this.userClosed || this.connectionFinalStatus != null)
+          && this.waitingRequestQueue.isEmpty();
+    } finally {
+      this.lock.unlock();
+    }
+  }
+
+  private void waitForDoneCallback() {
+    log.info("Waiting for done callback from stream connection. Stream: " + streamName);
+    while (true) {
+      this.lock.lock();
+      try {
+        if (connectionFinalStatus != null) {
+          // Done callback is received, return.
+          return;
+        }
+      } finally {
+        this.lock.unlock();
+      }
+      Uninterruptibles.sleepUninterruptibly(100, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void cleanupInflightRequests() {
+    Throwable finalStatus;
+    Deque<AppendRequestAndResponse> localQueue = new LinkedList<AppendRequestAndResponse>();
+    this.lock.lock();
+    try {
+      finalStatus = this.connectionFinalStatus;
+      while (!this.inflightRequestQueue.isEmpty()) {
+        localQueue.addLast(this.inflightRequestQueue.pollFirst());
+      }
+    } finally {
+      this.lock.unlock();
+    }
+    log.info(
+        "Cleaning "
+            + localQueue.size()
+            + " inflight requests with error: "
+            + finalStatus.toString());
+    while (!localQueue.isEmpty()) {
+      localQueue.pollFirst().appendResult.setException(finalStatus);
+    }
+  }
+
+  private void requestCallback(AppendRowsResponse response) {
+    AppendRequestAndResponse requestWrapper;
+    this.lock.lock();
+    try {
+      requestWrapper = this.inflightRequestQueue.pollFirst();
+    } finally {
+      this.lock.unlock();
+    }
+    requestWrapper.appendResult.set(response);
```

Comment:
For in stream error, should we also fail out as of now? It can be OUT_OF_RANGE error.

Reply:
We've stated clearly in the comment of append method that user should check the in stream error. It is the same in the old stream writer as well.

I would prefer to keep this layer as thin as possible to simplify the logic. I don't too much logic in the client library, which is hard to update. It seems unnecessary to me.

However, if you believe otherwise, we could discuss offline separately, and make separate CLs. if necessary.

We don't need to block this CL because of this. Can we move forward?

Reply:
Sure, there is nothing blocking.